packages/gatsby-plugin-utils/src/polyfill-remote-file/index.ts

```diff
@@ -0,0 +1,157 @@
+import {
+  generateGatsbyImageFieldConfig,
+  gatsbyImageResolver,
+} from "./graphql/gatsby-image-resolver"
+
+import type { Store } from "gatsby"
+import type { InterfaceTypeComposerAsObjectDefinition } from "graphql-compose"
+import type { SchemaBuilder, IRemoteFileNode } from "./types"
+
+let enums: ReturnType<typeof getRemoteFileEnums> | undefined
+
+export function getRemoteFileFields(
+  enums: ReturnType<typeof getRemoteFileEnums>,
+  store: Store
+): Record<string, unknown> {
+  return {
+    id: `ID!`,
+    mimeType: `String!`,
+    filename: `String!`,
+    filesize: `Int`,
+    width: `Int`,
+    height: `Int`,
+    publicUrl: generatePublicUrlFieldConfig(store),
+    resize: generateResizeFieldConfig(enums, store),
+    gatsbyImage: generateGatsbyImageFieldConfig(enums, store),
+  }
+}
+
+function addRemoteFilePolyfillInterface<
+  T = ReturnType<SchemaBuilder["buildObjectType"]>
+>(
+  type: T,
+  {
+    schema,
+    store,
+  }: {
+    schema: SchemaBuilder
+    store: Store
+  }
+): T {
+  // When the image-cdn is part of Gatsby we will only add the RemoteFile interface if necessary
+  if (hasFeature(`image-cdn`)) {
+    // @ts-ignore - wrong typing by typecomposer
+    if (!type.config.interfaces.includes(`RemoteFile`)) {
+      // @ts-ignore - wrong typing by typecomposer
+      type.config.interfaces.push(`RemoteFile`)
+    }
+
+    return type
+  }
+
+  if (!enums) {
+    // We only want to create the enums and interface once
+    const composer = new SchemaComposer()
+    enums = getRemoteFileEnums(composer.createEnumTC.bind(composer))
+
+    const types: Array<
+      | string
+      | ReturnType<SchemaBuilder["buildObjectType"]>
+      | ReturnType<SchemaBuilder["buildInterfaceType"]>
+      | ReturnType<SchemaBuilder["buildEnumType"]>
+    > = []
+
+    for (const key in enums) {
+      if (enums[key]) {
+        types.push(
+          schema.buildEnumType({
+            name: enums[key].getTypeName(),
+            values: enums[key].getFields(),
+          })
+        )
+      }
+    }
+
+    types.push(
+      schema.buildObjectType({
+        name: `RemoteFileResize`,
+        fields: {
+          width: `Int`,
+          height: `Int`,
+          src: `String`,
+        },
+      }),
+      schema.buildInterfaceType({
+        name: `RemoteFile`,
+        fields: getRemoteFileFields(
+          enums,
+          store
+        ) as InterfaceTypeComposerAsObjectDefinition<
+          IRemoteFileNode,
+          unknown
+        >["fields"],
+      })
+    )
+
+    // We need to use import-from to remove circular dependency
+    const actions = importFrom(
+      global.__GATSBY.root ?? process.cwd(),
+      `gatsby/dist/redux/actions`
+    )
+    store.dispatch(
+      // @ts-ignore - importFrom doesn't work with types
+      actions.createTypes(types, {
+        name: `gatsby`,
+        version: getGatsbyVersion(),
+        resolve: path.join(__dirname, `../`),
+      })
+    )
+  }
```

Comment:
I think we should try to avoid importing from gatsby internals.

Can we instead of passing `store` and importing actions, pass `actions` from source plugins and "just" use 2nd param on action to tell it's `gatsby` that should create a type - like in https://github.com/gatsbyjs/gatsby/blob/2cf474b7c5d85ab4c62379c872f57f2cc07bd876/packages/gatsby-source-filesystem/src/create-remote-file-node.js#L76-L80 for file node creation

Reply:
I'll see if I can get it to  work